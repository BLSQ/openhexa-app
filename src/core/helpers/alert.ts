export enum AlertType {
  success,
  error,
  info,
  warning,
}

export function displayAlert(message: string, type?: AlertType) {
  if (!window) {
    console.warn("displayAlert cannot be called on the server.");
    return;
  }
  const event = new CustomEvent("displayAlert", {
    detail: {
      message,
      type: type ?? AlertType.info,
    },
  });

  window.dispatchEvent(event);
}
