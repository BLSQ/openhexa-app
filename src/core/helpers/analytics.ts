import { NextRequest } from "next/server";
import getConfig from "next/config";

const { publicRuntimeConfig } = getConfig();

interface TrackEventProperties {
  [key: string]: any;
}

async function sendEvent(
  event: string,
  properties: TrackEventProperties,
  headers?: Headers,
): Promise<void> {
  if (publicRuntimeConfig.DISABLE_ANALYTICS === "true") {
    return;
  }
  const res = await fetch(
    `${publicRuntimeConfig.OPENHEXA_BACKEND_URL ?? ""}/analytics/track/`,
    {
      method: "POST",
      priority: "low",
      headers: headers,
      credentials: "include",
      body: JSON.stringify({ event, properties }),
    },
  );

  if (!res.ok) {
    throw new Error(`Failed to send event: ${res.statusText}`);
  }
}

export async function trackEvent(
  event: string,
  properties: TrackEventProperties,
): Promise<void> {
  if (typeof window === "undefined") {
    // We do not support tracking events on the server
    return;
  }
  try {
    await sendEvent(event, properties);
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}

export async function pageView(request: NextRequest) {
  // This is to be called on the server side only.
  if (typeof window !== "undefined") {
    return;
  }
  try {
    const headers = request.headers;
    headers.set("content-type", "application/json");

    await sendEvent(
      "pageview",
      {
        pathname: request.nextUrl.pathname,
        searchParams: request.nextUrl.searchParams.toString(),
      },
      headers,
    );
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}
