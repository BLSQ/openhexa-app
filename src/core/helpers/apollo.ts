import {
  ApolloClient,
  ApolloLink,
  InMemoryCache,
  NormalizedCacheObject,
  createHttpLink,
  InMemoryCacheConfig,
} from "@apollo/client";
import { onError } from "@apollo/link-error";
import merge from "deepmerge";
import { IncomingHttpHeaders } from "http";
import fetch from "isomorphic-unfetch";
import isEqual from "lodash/isEqual";
import type { AppProps } from "next/app";
import { useMemo } from "react";
import getConfig from "next/config";

const APOLLO_STATE_PROP_NAME = "__APOLLO_STATE__";

export type CustomApolloClient = ApolloClient<NormalizedCacheObject>;

let apolloClient: CustomApolloClient | undefined;

const { publicRuntimeConfig } = getConfig();

const CACHE_CONFIG: InMemoryCacheConfig = {
  // possibleTypes must be provided to cache correctly unions and interfaces
  // https://www.apollographql.com/docs/react/data/fragments/#using-fragments-with-unions-and-interfaces
  possibleTypes: {
    CollectionElement: [
      "DHIS2DataElementCollectionElement",
      "S3ObjectCollectionElement",
    ],
  },
  typePolicies: {
    Team: {
      merge: true,
      fields: {
        permissions: {
          merge: true,
        },
      },
    },
    DAGRun: {
      merge: true,
    },
    User: {
      merge: true,
    },
    Country: {
      // Country code are unique (at least it should). Let's use that for the cache key
      keyFields: ["code"],
    },
    CatalogEntryType: {
      merge: true,
    },
  },
};

const createApolloClient = (headers: IncomingHttpHeaders | null = null) => {
  const enhancedFetch = (url: RequestInfo, init: RequestInit) => {
    return fetch(url, {
      ...init,
      headers: {
        ...init.headers,
        cookie: headers?.cookie ?? "",
        accept: "application/json",
      },
    });
  };

  const link = ApolloLink.from([
    onError(({ graphQLErrors, networkError }) => {
      if (graphQLErrors) {
        graphQLErrors.forEach(({ message, locations, path }) =>
          console.error(
            `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
              locations
            )}, Path: ${path}`
          )
        );
      }
      if (networkError) {
        console.error(
          `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
        );
      }
    }),

    createHttpLink({
      uri: (operation) =>
        operation.operationName
          ? `${publicRuntimeConfig.GRAPHQL_ENDPOINT}${operation.operationName}/`
          : publicRuntimeConfig.GRAPHQL_ENDPOINT,
      fetch: enhancedFetch,
      credentials: "include",
      fetchOptions: {
        mode: "cors",
      },
    }),
  ]);

  const cache = new InMemoryCache(CACHE_CONFIG);

  return new ApolloClient({
    ssrMode: typeof window === "undefined",
    ssrForceFetchDelay: 100, // in milliseconds
    link,
    cache,
  });
};

type InitialState = NormalizedCacheObject | undefined;
export interface GetApolloClient {
  headers?: IncomingHttpHeaders | null;
  initialState?: InitialState | null;
}

export const getApolloClient = (
  { headers, initialState }: GetApolloClient = {
    headers: null,
    initialState: null,
  }
) => {
  const client = apolloClient ?? createApolloClient(headers);

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // get hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = client.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = merge(initialState, existingCache, {
      // combine arrays using object equality (like in sets)
      arrayMerge: (destinationArray, sourceArray) => [
        ...sourceArray,
        ...destinationArray.filter((d) =>
          sourceArray.every((s) => !isEqual(d, s))
        ),
      ],
    });
    // Restore the cache with the merged data
    client.cache.restore(data);
  }

  // For SSG and SSR always create a new Apollo Client
  if (typeof window === "undefined") {
    return client;
  } else if (!apolloClient) {
    // Create the Apollo Client once in the client
    apolloClient = client;
  }

  return client;
};

export const addApolloState = (client: ApolloClient<NormalizedCacheObject>) => {
  return { props: { [APOLLO_STATE_PROP_NAME]: client.cache.extract() } };
};

export function useApollo(
  pageProps: AppProps<{ [key: string]: any }>["pageProps"]
) {
  const state = pageProps[APOLLO_STATE_PROP_NAME];
  const client = useMemo(
    () => getApolloClient({ initialState: state }),
    [state]
  );
  return client;
}
