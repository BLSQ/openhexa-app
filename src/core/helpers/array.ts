export function ensureArray<T>(maybeArray: T[] | T): T[] {
  if (typeof maybeArray === "undefined") {
    return [];
  } else if (maybeArray === null) {
    return [];
  } else if (Array.isArray(maybeArray)) {
    return maybeArray;
  } else {
    return new Array(maybeArray);
  }
}
