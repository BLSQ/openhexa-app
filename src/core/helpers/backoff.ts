export type BackoffOptions = {
  min: number;
  max: number;
  factor: number;
  jitter: number;
};

class Backoff {
  private _attempts: number;
  private _base: number;
  private _max: number;
  private _factor: number;
  private _jitter: number;

  constructor(options: BackoffOptions) {
    this._base = options.min || 100;
    this._max = options.max || 10000;
    this._factor = options.factor || 2;
    this._jitter =
      options.jitter > 0 && options.jitter < 1 ? options.jitter : 0;
    this._attempts = 0;
  }

  get attempts(): number {
    return this._attempts;
  }

  duration() {
    let duration = this._base * Math.pow(this._factor, this._attempts++);
    if (this._jitter) {
      var rand = Math.random();
      var deviation = Math.floor(rand * this._jitter * duration);
      duration =
        (Math.floor(rand * 10) & 1) == 0
          ? duration - deviation
          : duration + deviation;
    }
    return Math.min(duration, this._max) | 0;
  }

  reset() {
    this._attempts = 0;
  }
}

export default Backoff;
