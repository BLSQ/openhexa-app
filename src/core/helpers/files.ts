import { createDeferred, Deferred } from "./promise";

type XHROptions = {
  url: string;
  method: "POST" | "PUT" | "PATCH";
  headers: Record<string, string>;
};

type ProgressFunc = (progress: number) => void;
type GetXHROptionsFunc<T extends File = File> = (
  file: T,
) => Promise<XHROptions>;

type JobOptions<T extends File> = {
  files: T[];
  getXHROptions: GetXHROptionsFunc<T>;
  onProgress?: ProgressFunc;
};
type JobItem<T extends File> = {
  job: Job<T>;
  deferred: Deferred;
};

/*
  A Job represents the upload of a set of files.
*/

export interface JobFile extends File {
  loaded?: number;
  isUploaded?: boolean;
}

class Job<T extends JobFile = JobFile> {
  private _files: T[];
  private getXHROptions: GetXHROptionsFunc<T>;
  private _onProgress?: ProgressFunc;
  private _runs: number = 0;
  private _status: "pending" | "running" | "done" | "error";
  private _error: undefined | Error;

  constructor({ files, getXHROptions, onProgress }: JobOptions<T>) {
    this._files = files;
    this.getXHROptions = getXHROptions;
    this._onProgress = onProgress;
    this._status = "pending";
  }

  public get progress() {
    let total = 0;
    let loaded = 0;
    for (const file of this._files) {
      if (file.isUploaded) {
        loaded += file.size;
      }
      total += file.size;
    }
    return Math.round((loaded * 100) / (total || 1));
  }

  public get runs() {
    return this._runs;
  }

  public get status() {
    return this._status;
  }

  public get error() {
    return this._error;
  }

  onProgress() {
    if (this._onProgress) {
      try {
        this._onProgress(this.progress);
      } catch (err) {
        console.error("Error in provided onProgress function.", err);
      }
    }
  }

  async run(): Promise<any> {
    this._runs++;
    this._status = "running";

    for (const file of this._files) {
      if (file.isUploaded) continue; // In case of retry, do not upload again the files that were already uploaded

      console.log(`Uploading "${file.name}"`);

      // Get Axios request options
      const options = await this.getXHROptions(file);

      console.log(`Start upload of "${file.name}"`, options);
      try {
        const xhr = new XMLHttpRequest();
        const success = await new Promise((resolve) => {
          // Add event listeners to track the upload progress
          xhr.upload.addEventListener("progress", (event) => {
            if (event.lengthComputable) {
              file.loaded = event.loaded;
              this.onProgress();
            }
          });

          xhr.addEventListener("loadend", () => {
            resolve(xhr.readyState === 4 && xhr.status === 200);
          });

          xhr.open(options.method, options.url, true);

          for (const key in options.headers) {
            xhr.setRequestHeader(key, options.headers[key]);
          }

          xhr.send(file);
        });
        if (success) {
          console.log(`Upload of "${file.name}" completed.`);
          file.isUploaded = true;
        }
      } catch (err: any) {
        this._error = err;
        this._status = "error";
        throw err;
      }
    }

    this._status = "done";

    return;
  }
}

/*
  The UploadManager is the main part of the upload mechanism. It keeps the list of all upload jobs in the app.
  User can create new upload job with: 
  
    manager.createUploadJob({
      files: [],
      getXHROptions: () => {} // Returns the required options to execute the request
    })
  
*/
class UploadManager<T extends JobFile = JobFile> {
  pendingJobs: JobItem<T>[];
  runningJobs: JobItem<T>[];
  maxConcurrentJobs: number;

  constructor(maxConcurrentJobs: number = 2) {
    this.maxConcurrentJobs = maxConcurrentJobs;
    this.pendingJobs = [];
    this.runningJobs = [];
  }

  removeRunningJob(job: Job<T>) {
    this.runningJobs = this.runningJobs.filter((item) => item.job !== job);
  }

  process() {
    if (this.runningJobs.length >= this.maxConcurrentJobs) {
      // We already run the maximum jobs concurrently
      return;
    }

    // Try to get the first pending job of the list
    const next = this.pendingJobs.shift();
    if (!next) {
      return;
    }
    this.runningJobs.push(next);
    const { job, deferred } = next;

    job.run().then(
      (result) => {
        this.removeRunningJob(job);
        deferred.resolve(result);
        this.process();
      },
      (err) => {
        console.log("onJobFailure", job, err);
        this.removeRunningJob(job);
        if (job.runs >= 3) {
          err.job = job;
          deferred.reject(err);
        } else {
          setTimeout(
            () => {
              this.pendingJobs.push(next);
              this.process();
            },
            2 ^ (job.runs + 1),
          );
        }
        this.process();
      },
    );
  }

  createUploadJob(options: JobOptions<T>) {
    const job = new Job<T>(options);
    const deferred = createDeferred();

    this.pendingJobs.push({ job, deferred });
    this.process();

    return deferred.promise;
  }
}

export const uploader = new UploadManager();
