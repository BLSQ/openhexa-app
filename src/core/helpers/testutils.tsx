import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { faker } from "@faker-js/faker";
import { ByRoleOptions, screen } from "@testing-library/react";
import { ReactNode } from "react";

type TestAppProps = {
  children: ReactNode;
  mocks?: MockedResponse[];
};

jest.mock("next/router", () => require("next-router-mock"));
// This is needed for mocking 'next/link':
jest.mock("next/dist/client/router", () => require("next-router-mock"));
jest.mock("identity/hooks/useMe");

export function TestApp(props: TestAppProps) {
  return (
    <MockedProvider addTypename={false} mocks={props.mocks ?? []}>
      {props.children}
    </MockedProvider>
  );
}

export const LOGGED_IN_USER = {
  id: faker.datatype.uuid(),
  firstName: faker.name.firstName,
  lastName: faker.name.lastName,
  avatar: {
    initials: "AE",
    color: "red",
  },
};

export async function waitForDialog(options?: ByRoleOptions) {
  const dialog = await screen.getByRole("dialog", options);
  return dialog;
}
