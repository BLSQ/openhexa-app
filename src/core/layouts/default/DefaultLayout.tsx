import { CustomApolloClient } from "core/helpers/apollo";
import type { ReactElement } from "react";
import Header from "./Header";
import PageContent from "./PageContent";

type LayoutProps = {
  children: ReactElement;
  pageProps: any;
};

const Layout = (props: LayoutProps) => {
  const { children } = props;
  return (
    <div className="flex min-h-screen flex-col ">
      <Header />
      {children}
    </div>
  );
};

Layout.PageContent = PageContent;

Layout.prefetch = async (client: CustomApolloClient) => {};

export default Layout;
