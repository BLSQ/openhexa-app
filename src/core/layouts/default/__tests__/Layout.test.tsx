import { render, screen } from "@testing-library/react";
import { TestApp } from "core/helpers/testutils";
import mockRouter from "next-router-mock";
import DefaultLayout from "../DefaultLayout";

describe("DefaultLayout", () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl("/");
  });

  it("renders the layout for an unauthenticated user", async () => {
    const { container, debug } = render(
      <TestApp me={{ user: null, features: [] }}>
        <DefaultLayout pageProps={{}}>
          <span data-testid="page" />
        </DefaultLayout>
      </TestApp>,
    );

    expect(screen.getByTestId("page")).toBeInTheDocument();
    expect(screen.queryByText("Dashboard")).toBeNull();

    expect(container).toMatchSnapshot();
  });

  it("renders the layout for a authenticated user", async () => {
    const { container, debug } = render(
      <TestApp>
        <DefaultLayout pageProps={{}}>
          <span data-testid="page" />
        </DefaultLayout>
      </TestApp>,
    );
    expect(screen.getByText("Catalog")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
