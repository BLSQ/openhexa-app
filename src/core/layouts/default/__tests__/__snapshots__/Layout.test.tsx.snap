// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultLayout renders the layout for a authenticated user 1`] = `
<div>
  <div
    class="flex min-h-screen flex-col "
  >
    <div
      class="bg-gray-800"
    >
      <div
        class="w-full max-w-7xl mx-auto px-2 sm:px-6 lg:px-8 flex h-16 items-center gap-6"
      >
        <a
          class="relative flex h-8 flex-shrink-0 items-center"
          href="/"
        >
          <img
            alt="OpenHexa logo"
            class="block h-full lg:hidden"
            src="/images/logo.svg"
          />
          <img
            alt="OpenHexa logo"
            class="hidden h-full lg:block"
            src="/images/logo_with_text_white.svg"
          />
        </a>
        <nav
          class="relative flex flex-1 items-center space-x-4 text-sm font-medium"
        >
          <a
            class="px-3 py-2 rounded-md hover:bg-gray-700 hover:text-white bg-gray-900 text-white"
            href="/"
          >
            Dashboard
          </a>
          <a
            class="px-3 py-2 rounded-md hover:bg-gray-700 hover:text-white text-gray-300"
            href="/catalog"
          >
            Catalog
          </a>
          <a
            class="px-3 py-2 rounded-md hover:bg-gray-700 hover:text-white text-gray-300"
            href="/notebooks"
          >
            Notebooks
          </a>
          <a
            class="px-3 py-2 rounded-md hover:bg-gray-700 hover:text-white text-gray-300"
            href="/pipelines"
          >
            Pipelines
          </a>
        </nav>
        <button
          class="group mr-4 hidden flex-shrink cursor-pointer items-center space-x-2 text-gray-400 hover:text-white md:inline-flex"
        >
          <svg
            aria-hidden="true"
            class="h-5"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <span>
            Search
          </span>
          <span
            class="inline-flex items-center rounded border border-gray-400 p-1 text-xs font-medium shadow-sm focus:outline-none group-hover:border-white"
          >
            âŒ˜K
          </span>
        </button>
        <div
          class="relative"
          data-headlessui-state=""
        >
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            data-headlessui-state=""
            id="headlessui-menu-button-6fbe024f-2316-4265-a6e8-d65a837e308a"
            type="button"
          >
            <span
              class="inline-flex flex-shrink-0 items-center justify-center rounded-full bg-gray-500 text-md h-10 w-10"
              style="background: red;"
            >
              <span
                class="font-medium leading-none text-white"
              >
                AE
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <span
      data-testid="page"
    />
  </div>
</div>
`;

exports[`DefaultLayout renders the layout for an unauthenticated user 1`] = `
<div>
  <div
    class="flex min-h-screen flex-col "
  >
    <span
      data-testid="page"
    />
  </div>
</div>
`;
